cluster-autoscaler/cloudprovider/digitalocean/digitalocean_manager.go

```diff
@@ -0,0 +1,248 @@
+	})
+	oauthClient := oauth2.NewClient(context.Background(), tokenSource)
+
+	opts := []godo.ClientOpt{}
+	if cfg.URL != "" {
+		opts = append(opts, godo.SetBaseURL(cfg.URL))
+	}
+
+	opts = append(opts, godo.SetUserAgent("cluster-autoscaler-digitalocean/"+version))
+
+	doClient, err := godo.New(oauthClient, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't initialize DigitalOcean client: %s", err)
+	}
+
+	m := &Manager{
+		client:     doClient.Kubernetes,
+		clusterID:  cfg.ClusterID,
+		nodeGroups: make([]*NodeGroup, 0),
+	}
+
+	return m, nil
+}
+
+// Refresh refreshes the cache holding the nodegroups. This is called by the CA
+// based on the `--scan-interval`. By default it's 10 seconds.
+func (m *Manager) Refresh() error {
+	ctx := context.Background()
+	nodePools, _, err := m.client.ListNodePools(ctx, m.clusterID, nil)
+	if err != nil {
+		return err
+	}
+
+	var group []*NodeGroup
+	for _, np := range nodePools {
+		nodePool, resp, err := m.client.GetNodePool(ctx, m.clusterID, np.ID)
+		if err != nil {
+			if resp != nil && resp.StatusCode == http.StatusNotFound {
+				return ErrNodePoolNotExist
+			}
+
+			return err
+		}
+
+		spec, err := parseTags(nodePool.Tags)
+		if err != nil {
+			// we should not return an error here, because one misconfigured
+			// node pool shouldn't bring down the whole cluster-autoscaler
+			klog.V(4).Infof("skipping misconfigured node pool: %q name: %s tags: %+v err: %s",
+				nodePool.ID, nodePool.Name, nodePool.Tags, err)
+			continue
+		}
+
+		if !spec.enabled {
+			continue
+		}
+
+		minSize := minNodePoolSize
+		if spec.min != 0 {
+			minSize = spec.min
+		}
+
+		maxSize := maxNodePoolSize
+		if spec.max != 0 {
+			maxSize = spec.max
+		}
+
+		klog.V(4).Infof("adding node pool: %q name: %s min: %d max: %d",
+			nodePool.ID, nodePool.Name, minSize, maxSize)
+
+		group = append(group, &NodeGroup{
+			id:        np.ID,
+			clusterID: m.clusterID,
+			client:    m.client,
+			nodePool:  nodePool,
+			minSize:   minSize,
+			maxSize:   maxSize,
+		})
+	}
+
+	if len(group) == 0 {
+		klog.V(4).Info("cluster-autoscaler is disabled. no node pools are configured")
+	}
+
+	m.nodeGroups = group
+	return nil
+}
+
+// nodeSpec defines a custom specification for a given node
+type nodeSpec struct {
+	min     int
+	max     int
+	enabled bool
+}
+
+// parseTags parses a list of tags from a DigitalOcean node pool
+func parseTags(tags []string) (*nodeSpec, error) {
+	spec := &nodeSpec{}
+
+	for _, tag := range tags {
+		if !strings.HasPrefix(tag, tagPrefix) {
+			continue
+		}
+
+		splitted := strings.Split(strings.TrimPrefix(tag, tagPrefix), ":")
+		if len(splitted) != 2 {
+			return nil, fmt.Errorf("malformed tag: %q", tag)
+		}
+
+		key, value := splitted[0], splitted[1]
+
+		switch key {
+		case "enabled":
+			if value == "true" {
+				spec.enabled = true
+			}
+		case "min":
+			min, err := strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("invalid minimum nodes: %q", value)
+			}
```

Comment:
Explicitly return an error for min=0? Currently you just silently override it to 1. Many other providers support scale-to-0 so it's possible for someone to assume digitalocean also supports it and just set it.

Reply:
We don't support `0` yet :/ Hence I return an error here. This is one of our short comings on our end we're working on fixing it.

Reply:
My comment is that you don't actually return any error. You just treat 0 as 'not set' and silently default to 1.

Reply:
Interestingly it seems like you can also set max to a negative value as long as you don't set min?

Reply:
Yeah you're right. Added this test case and it fails:

```
		{
			name: "bad tags - max is set to negative, no min",
			tags: []string{
				"k8s-cluster-autoscaler-enabled:true",
				"k8s-cluster-autoscaler-max:-5",
			},
			wantErr: true,
		},
```

Making sure to fix it.